liberum_vault/src/vault.rs

```diff
@@ -0,0 +1,481 @@
+        let fragment_sizes = Self::fragment_sizes(file_size);
+        let mut current_pos = 0;
+        let mut result: Vec<FragmentData> = Vec::new();
+
+        for fragment_size in fragment_sizes {
+            let mut f = File::open(path).await?;
+            f.seek(std::io::SeekFrom::Start(current_pos)).await?;
+
+            let buf_reader = BufReader::new(f);
+            let reader_stream = ReaderStream::new(buf_reader.take(fragment_size));
+
+            result.push(reader_stream.boxed());
+            current_pos += fragment_size;
+        }
+
+        Ok(result)
+    }
+
+    fn fragment_sizes(target: u64) -> Vec<u64> {
+        let mut fragment_sizes = Vec::new();
+        let mut current_size = cmp::max(Self::MIN_FRAGMENT_SIZE, Self::power_2_upto(target));
+        let mut current_target = target;
+
+        while current_target != 0 {
+            current_size = cmp::max(
+                Self::MIN_FRAGMENT_SIZE,
+                Self::power_2_desc_from_power_2(current_size, current_target),
+            );
+
+            fragment_sizes.push(current_size);
+            current_target = current_target.saturating_sub(current_size);
+        }
+
+        fragment_sizes
+    }
+
+    fn power_2_upto(limit: u64) -> u64 {
+        let powers_from_1 = successors(Some(1u64), |&n| Some(n * 2))
+            .take_while(|x| x <= &limit)
+            .last();
+
+        once(0u64).chain(powers_from_1).last().unwrap()
+    }
+
+    fn power_2_desc_from_power_2(start: u64, limit: u64) -> u64 {
+        assert!(Self::is_power_of_2(start));
+
+        if limit == 0 {
+            return 0;
+        }
+
+        successors(Some(start), |&n| Some(n / 2))
+            .take_while(|x| x > &(limit / 2) || x == &limit)
+            .last()
+            .unwrap()
+    }
+
+    fn is_power_of_2(number: u64) -> bool {
+        number > 0 && (((number) & (number - 1)) == 0)
+    }
+
+    async fn prepare_db(&self) -> Result<()> {
+        const CREATE_FRAGMENT_TABLE_QUERY: &str = "
+            CREATE TABLE IF NOT EXISTS fragment (
+                hash0 INTEGER,
+                hash1 INTEGER,
+                hash2 INTEGER,
+                hash3 INTEGER,
+                path VARCHAR(255),
+                size INTEGER,
+                PRIMARY KEY (hash0, hash1, hash2, hash3)
+            )
+        ";
+
+        self.db
+            .call(|conn| Ok(conn.execute(CREATE_FRAGMENT_TABLE_QUERY, ())?))
+            .await?;
+
+        Ok(())
+    }
+
+    async fn load_fragment_info(&self, key: Key) -> Result<Option<FragmentInfo>> {
+        const SELECT_FRAGMENT_QUERY: &str = "
+            SELECT path, size
+            FROM fragment
+            WHERE hash0 = ?1 AND hash1 = ?2 AND hash2 = ?3 AND hash3 = ?4
+        ";
+
+        self.db
+            .call(move |conn| {
+                let mut stmt = conn.prepare(SELECT_FRAGMENT_QUERY)?;
+                let key_u64_slice = key.as_u64_slice_be();
```

Comment:
Big Endian?

Reply:
Yes, this does not really matter that much as long as we are reading and saving to DB using the same endianess.